### PR TITLE
Feature edit delete

### PR DESCRIPTION
**Pull request by agonzalez0515**

### CODE REVIEW

File: src/main/java/webserver/Router.java [hunk 1]
```diff
@@ -58,6 +68,10 @@ public String route(Request request) {
                 return this.HEADRoutes;
             case HTTP_HEADERS.POST:
                 return this.POSTRoutes;
+            case HTTP_HEADERS.PUT:
```
Comment (kkrull) [inline src/main/java/webserver/Router.java]: There's an opportunity to use polymorphism here.  Namely, you could call something like `method.getRoutes()`.  So the `PUT` method would return `PUTRoutes`, `GET` method would return `GETRoutes`, etc...

This would require converting `String method` to a value type like `class Method {...}` or `enum Method {...}.

That said, I'm not recommending a change here; only pausing for a moment to reflect.  While it's generally said that "switch is not object-oriented", that doesn't mean that all uses of `switch` in a language like Java are inappropriate.  Whether to use `switch` or polymorphism really comes down to these questions

> How many places do I need to modify, when adding a new HTTP method?
> How likely am I to need to add a new HTTP method?

In your case, the HTTP methods are pretty well fixed and unlikely to change and I think this is the only place where I've seen the `switch` logic.  It's not like you have to modify a dozen classes (which would be a violation of the Open-Closed Principle) when adding a new HTTP method.  I forget where I saw this; Robert Martin summarized this really well in one of his books or videos, but I'm not finding a link to it right now.

TL;DR - no change required; just wanted to inspire some reflection.

File: src/test/java/webserver/data/JsonClientTest.java [hunk 2]
```diff
@@ -1,12 +1,14 @@
-package webserver.data;
+                              package webserver.data;
```
Comment (kkrull) [inline src/test/java/webserver/data/JsonClientTest.java]: Fell asleep on the keyboard? _he says, from experience_

File: src/test/java/webserver/data/JsonClientTest.java [hunk 3]
```diff
@@ -18,10 +20,18 @@
 @SuppressWarnings("unchecked")
 public class JsonClientTest {
 
-    //TODO setup a before to seed the json and an after to clear the json file. 
+    @Before
+    public void setUpData() {
+        TestHelper.copyFile("src/test/resources/test.json", "src/test/resources/todos.json");
```
Comment (kkrull) [inline src/test/java/webserver/data/JsonClientTest.java]: Looks like a good use of a test fixture - nice work!

File: src/test/java/webserver/helpers/TestHelper.java [hunk 4]
```diff
@@ -0,0 +1,48 @@
...
+import java.io.IOException;
+
+public class TestHelper {
+
+    public static void copyFile(String copyFromPath, String copyToPath) {
+        FileInputStream in = null;
+        FileOutputStream out = null;
+
+        try {
+            File inFile = new File(copyFromPath);
```
Comment (kkrull) [inline src/test/java/webserver/helpers/TestHelper.java]: The general algorithm looks fine. Here's a good place to use try-with-resources

```java
try(FileInputStream in = new FileInputStream(inFile);
    FileInputStream out = new FileInputStream(outFile)) {
  ...read from in and write to out...
}
//no need to call #close here; it's called automatically by try-with-resources
```

File: src/test/java/webserver/data/JsonClientTest.java (hunk 2 above)
Comment (agonzalez0515) [reply src/test/java/webserver/data/JsonClientTest.java]: haha. oops. no idea!

File: src/main/java/webserver/controllers/TodoController.java [hunk 5]
```diff
@@ -98,10 +99,70 @@ public TodoController(TodoService todoService) {
...
     };
 
+    public Callback<Request, String> getTodoDetailEdit = (request) -> {
+        final int id = getTodoIdFromPath(request.getPath());
+        String html = "";
+        int responseCode = 0;
+
+        Mustache template = MustacheUtil.getTemplate("todo-item-edit.mustache");
+        List<Todo> allTodos = todoService.getTodos();
+        Todo todo = TodosHelper.getTodo(id, allTodos);
```
Comment (kkrull) [inline src/main/java/webserver/controllers/TodoController.java]: This could be a good place to introduce a method to your service/interface for TODO items.  Something like 

```java
interface TodoService {
  Todo getTodo(int id);
}
```

The implementation will probably have the same two lines of code in it, but there's real power in expressing _exactly_ what you want when looking up data.  It can also make mocking/stubbing easier in your tests.

This comes up in data layers a lot, where the database might be querying way more records than necessary (looking up all todo items, for example) only for the application to sort through them and throw most of those results out.

File: src/main/java/webserver/controllers/TodoController.java [hunk 6]
```diff
@@ -98,10 +99,70 @@ public TodoController(TodoService todoService) {
...
+
+        if (todo == null) {
```
Comment (kkrull) [inline src/main/java/webserver/controllers/TodoController.java]: Here's a good place to use

```java
Optional.ofNullable(todo)
  .map(todo -> MustacheUtil.executeTemplate(template, todo))
  .orElseGet(() -> getTodoNotFound.apply(request));
```

Comment (kkrull) [reply src/main/java/webserver/controllers/TodoController.java]: ^^ and that's where returning `Optional` from `TodoService` can come in handy, since you're probably going to do some sort of null/empty check anyway, after you call it.

File: src/main/java/webserver/controllers/TodoController.java [hunk 7]
```diff
@@ -98,10 +99,70 @@ public TodoController(TodoService todoService) {
...
+    };
+
+    public Callback<Request, String> editTodo = (request) -> {
+        int id = getTodoIdFromPath(request.getPath());
+        String contentType = request.getHeaders().get(HTTP_HEADERS.CONTENT_TYPE);
+        String body = request.getBody();
+        List<Todo> allTodos = todoService.getTodos();
+        Todo todo = TodosHelper.getTodo(id, allTodos);
+ 
+        if (contentType.equals("application/x-www-form-urlencoded")) {
```
Comment (kkrull) [inline src/main/java/webserver/controllers/TodoController.java]: Pro-tip: It can come in handy to swap the order of the two strings.  `contentType` could be null if the header isn't set at all, which would cause `contentType.equals` to blow up.  But the reverse won't ever cause a null pointer exception, since it's a String constant

```java
if("application/x-www-form-urlencoded".equals(contentType)) ...
```

File: src/main/java/webserver/controllers/TodoController.java [hunk 8]
```diff
@@ -147,13 +208,30 @@ public TodoController(TodoService todoService) {
...
+            .withHeader(HTTP_HEADERS.CONTENT_TYPE, "text/html; charset=utf-8")
+            .build();
+    };
+
     private String getContentLength(String body) {
         return Integer.toString(body.length());
     }
  
-    private int getTodoIdFromPath(String path) {
+    private Integer getTodoIdFromPath(String path) {
```
Comment (kkrull) [inline src/main/java/webserver/controllers/TodoController.java]: Good use of using capital-i `Integer` (which can be null, in a meaningful way) over lowercase-i `int`.  I used to do this a lot too, before `Optional` came along.

File: src/main/java/webserver/data/JsonClient.java [hunk 9]
```diff
@@ -75,4 +75,21 @@ public void updateItemById(int id, JSONObject updatedItem) {
...
+        try (FileWriter file = fileUtils.fileWriter(todoFilePath)) {
+            for (int i = 0; i < todos.size(); i++) {
+                JSONObject obj = (JSONObject) todos.get(i);
+                int idToMatch = Integer.parseInt(obj.get("id").toString());
+                if (id == idToMatch) {
+                    todos.remove(i);
+                }
+            }
+            file.write(todos.toJSONString()); 
+            file.flush();
```
Comment (kkrull) [inline src/main/java/webserver/data/JsonClient.java]: Doesn't hurt, but you probably don't need to flush the stream.  This will be done automatically when you close it (which happens with you exit the try-with-resource block).

**Review by kkrull — APPROVED** [review]
This project is really taking shape - good work!  

Notice how - now that you have done a few PRs on this codebase and some refactoring - adding these new features doesn't affect so many files or result in that many new lines of code.  This is what you start to see as your design becomes more SOLID: something that sounds like it should be simple really is that simple, as expressed in the code.  That doesn't happen automatically though.  It's the result of your diligence in refactoring towards the SOLID principles.
